src/lib/mina_net2/bitswap_block.ml

```diff
@@ -0,0 +1,237 @@
+
+let absolute_max_links_per_block = Stdint.Uint16.(to_int max_int)
+
+type schema =
+  { total_blocks : int
+  ; full_link_blocks : int
+  ; non_max_link_block_count : int
+  ; last_block_data_size : int
+  ; max_block_size : int
+  ; max_links_per_block : int
+  }
+[@@deriving compare, eq, sexp]
+
+let required_bitswap_block_count ~max_block_size data_length =
+  if data_length <= max_block_size - 2 then 1
+  else
+    let n1 = data_length - link_size in
+    let n2 = max_block_size - link_size - 2 in
+    (n1 / n2) + if n1 mod n2 > 0 then 1 else 0
+
+let max_links_per_block ~max_block_size =
+  let links_per_block = (max_block_size - 2) / link_size in
+  min links_per_block absolute_max_links_per_block
+
+let create_schema ~max_block_size data_length =
+  let total_blocks = required_bitswap_block_count ~max_block_size data_length in
+  let last_block_data_size =
+    data_length - ((max_block_size - link_size - 2) * (total_blocks - 1))
+  in
+  let max_links_per_block = max_links_per_block ~max_block_size in
+  let non_max_link_block_count = (total_blocks - 1) mod max_links_per_block in
+  let full_link_blocks = (total_blocks - 1) / max_links_per_block in
+  { total_blocks
+  ; last_block_data_size
+  ; full_link_blocks
+  ; non_max_link_block_count
+  ; max_block_size
+  ; max_links_per_block
+  }
+
+let create_schema_length_prefixed ~max_block_size data_length =
+  create_schema ~max_block_size (data_length + 4)
+
+let blocks_of_data ~max_block_size data =
+  if max_block_size <= 2 + link_size then failwith "Max block size too small" ;
+  let max_data_chunk_size = max_block_size - 2 in
+  let data_length = Bigstring.length data in
+  let schema = create_schema ~max_block_size data_length in
+  let remaining_data = ref data_length in
+  let blocks = Blake2.Table.create () in
+  let link_queue = Queue.create () in
+  let dequeue_chunk chunk_size =
+    assert (!remaining_data >= chunk_size) ;
+    let chunk =
+      Bigstring.sub_shared data
+        ~pos:(!remaining_data - chunk_size)
+        ~len:chunk_size
+    in
+    remaining_data := !remaining_data - chunk_size ;
+    chunk
+  in
+  let dequeue_links num_links =
+    assert (Queue.length link_queue >= num_links) ;
+    let links = ref [] in
+    for _ = 1 to num_links do
+      links := Queue.dequeue_exn link_queue :: !links
+    done ;
+    !links
+  in
+  let create_block links chunk_size =
+    let chunk = dequeue_chunk chunk_size in
+    let num_links = List.length links in
+    let size = 2 + (num_links * link_size) + chunk_size in
+    if num_links > absolute_max_links_per_block || size > max_block_size then
+      failwith "invalid block produced" ;
+    let block = Bigstring.create size in
+    Bigstring.set_uint16_le_exn block ~pos:0 num_links ;
+    List.iteri links ~f:(fun i link ->
+        let link_buf = Bigstring.of_string (Blake2.to_raw_string link) in
+        Bigstring.blit ~src:link_buf ~src_pos:0 ~dst:block
+          ~dst_pos:(2 + (i * link_size))
```

Comment:
I'm reading the chunks in the opposite order that you are (reading chunks from the right to left instead of left to right). So by not reversing the indices, I believe I end up with the same order of linked data. Writing a test to compare the bitswap and ocaml versions, as you suggested, should shed light on whether or not I did this correctly.

Reply:
Actually, I'm also dequeuing links in the opposite order that you are. So perhaps this is still the wrong order. Either way, implementing a test seems like the best thing to do here to be totally sure.